test_cases/stdlib/builtins/test_dict.py

```diff
@@ -0,0 +1,40 @@
+# pyright: reportUnnecessaryTypeIgnoreComment=true
+
+from typing import Dict, Generic, Iterable, Tuple, TypeVar
+from typing_extensions import assert_type
+
+# These do follow `__init__` overloads order:
+bad = dict()  # type: ignore
+good: Dict[str, str] = dict()
+assert_type(good, Dict[str, str])
+
+assert_type(dict(arg=1), Dict[str, int])
+
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+
+
+class KeysAndGetItem(Generic[_KT, _VT]):
+    def keys(self) -> Iterable[_KT]:
+        ...
+
+    def __getitem__(self, __k: _KT) -> _VT:
+        ...
+
+
+kt1: KeysAndGetItem[int, str] = KeysAndGetItem()
+assert_type(dict(kt1), Dict[int, str])
+dict(kt1, arg="a")  # type: ignore
+
+kt2: KeysAndGetItem[str, int] = KeysAndGetItem()
+assert_type(dict(kt2, arg=1), Dict[str, int])
+
+i1: Iterable[Tuple[int, str]] = [(1, "a"), (2, "b")]
+assert_type(dict(i1), Dict[int, str])
```

Comment:
To work around the mypy/pyright disagreement here:

```suggestion
d = dict(i1)
assert_type(next(iter(d.keys())), int)
assert_type(d[1], str)
```

Reply:
I tried `if MYPY:` hack, it didn't work 😞